File: primo_casado.cpp
#include <stdio.h> 
#include <math.h>

/* Se n e n+2 são números primos, então dizem-se primos casados. 
 * Escreva um programa usando funções que receba um inteiro e determine se este é: 
 * "NAO PRIMO!", "PRIMO!" ou "PRIMO CASADO!".
 * */

bool eh_primo(int num) {
	if (num % 2 == 0)
		return false;
	for(int i = 3; i <= sqrt(num); i += 2) {
		if (num % i == 0)
			return false;
	}
	return true;
}

int main() {
	
	int num;
	
	scanf("%d", &num);
	
	if(eh_primo(num)) {
		if (eh_primo(num + 2))
			printf("PRIMO CASADO!\n");
		else
			printf("PRIMO!\n");
	} else {
		printf("NAO PRIMO!\n");
	}
	
	return 0;
	
}

File: praticando_programacao_funcional_parte_1/primo_casado.hs
{- Se n e n+2 são números primos, então dizem-se primos casados. 
   Escreva um programa usando funções que receba um inteiro e determine se este é: 
   "NAO PRIMO!", "PRIMO!" ou "PRIMO CASADO!".
 -}

-- mesmo algoritmo da questão anterior
divisoresDeNumero :: Int -> [Int]
divisoresDeNumero num = 
    [ x | x <- [1 .. metadeNumero], ((mod num x) == 0)]
    where metadeNumero = floor(fromIntegral num / 2)

ehPrimo :: Int -> Bool
ehPrimo n = if (divisoresDeNumero n) == [1] then True else False

ehCasado :: Int -> String
ehCasado n
    | ehPrimo(nSoma) && ehPrimo(n) = "PRIMO CASADO!"
    | ehPrimo(n) = "PRIMO!"
    | otherwise = "NAO PRIMO!"
    where nSoma = n + 2

main = do
    valor1 <- getLine
    let result = ehCasado (read valor1)
    print result
    
